### PR TITLE
cosmwasm-book-2.0

### PR DESCRIPTION
**Pull request by costa2400**
I revised the overall structure of the CosmWasm documentation to incorporate new features and optimizations available in version 2.0. This restructuring involved categorizing sections to delineate between basic concepts, advanced features, specialized topics, and real-world applications of CosmWasm.

Still WIP

### CODE REVIEW

File: src/part2/interacting-with-contracts.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+/// Maps an address to a balance
+/// You can use this to store a balance per address and load it later
+const BALANCE: Map<Addr, Uint128> = Map::new("balance");
+```
+
+### Sending and Querying Data
+
+Interacting with contracts involves sending transactions and querying data. This section provides detailed guidance on these operations.
+
+### Sending Transactions
```
Comment (chipshort) [inline src/part2/interacting-with-contracts.md:35]: Is this section about interacting from outside of a contract or sending messages to another contract from inside a contract?
- If it's about interacting from the outside: I don't know if there is a Rust framework for doing that. There's cosmjs for doing it in Typescript or you can use the chain binary.
- If it's about contract to contract messages: Those are not called "transactions". They are called "messages". A transaction can contain multiple messages. Inside the contract, you can add messages to the transaction, but not a new transaction.

I added some example code for contract to contract interactions here.

Comment (costa2400) [reply src/part2/interacting-with-contracts.md:35]: Thank you, @chipshort, for your feedback. I've updated the documentation to clarify external versus internal contract interactions and corrected the terminology to "messages" for contract-to-contract communication. Here's a summary:

External Interactions: Explained how to interact with contracts using CosmJS or CLI tools.
Contract-to-Contract Communications: Revised to highlight message-passing within CosmWasm.

File: src/introduction/what's-new-2.0.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+The transition from CosmWasm 1.5 to 2.0 marks a significant leap in the framework's evolution. This upgrade responds to the growing needs of the developer community and the ever-changing landscape of blockchain technology.
+
+## Changelog Highlights
+
+For a detailed list of all changes, improvements, and bug fixes in CosmWasm 2.0, we encourage you to review the official changelog on [GitHub](https://github.com/CosmWasm/cosmwasm/blob/main/CHANGELOG.md). The changelog provides an exhaustive breakdown of the updates, offering insights into the development focus areas and enhancements that have been prioritized in this release.
+
+## Highlights from the changelog include:
+
+- **IBC Integration**: Enhanced support for Inter-Blockchain Communication (IBC), allowing CosmWasm contracts to seamlessly interact with other IBC-enabled blockchains.
+- **Performance Optimizations**: Significant improvements in contract execution speed and efficiency, reducing gas costs and improving the overall user experience.
```
Comment (chipshort) [inline src/introduction/what's-new-2.0.md:16]: We haven't really done any changes to IBC or performance

File: src/introduction/what's-new-2.0.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+- **Security Enhancements**: Introduction of new security protocols and measures to strengthen contract integrity and protect against emerging vulnerabilities.
```
Comment (chipshort) [inline src/introduction/what's-new-2.0.md:17]: The only "Security Enhancement" we did is this: https://github.com/CosmWasm/cosmwasm/pull/1991/files. It's an improved fix for one very specific vulnerability we had, but that's about it. All the other supported versions are also fixed, just not with the improved fix.

File: src/introduction/what's-new-2.0.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+- **Developer Tooling** : Upgrades to the developer toolkit, including better debugging and testing capabilities, to streamline the smart contract development process.
```
Comment (chipshort) [inline src/introduction/what's-new-2.0.md:18]: We also didn't really do dev tooling changes. The only thing we improved with regard to debugging is the [backtraces](https://github.com/CosmWasm/cosmwasm/blob/7533a213c3dd171328c0e4f50a72d7daeff52fe8/CHANGELOG.md?plain=1#L80-L82)

File: src/part2/smart-contract-development.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    .unwrap();
+assert_eq!(key, b"1");
+assert_eq!(d, maria_williams);
+```
+
+### Example Integrating External Data
+
+```
+// Example Rust code for integrating external data sources
+// Placeholder for external data integration example
```
[the remark this answers is not in the document]
Comment (costa2400) [reply src/part2/smart-contract-development.md:98]: Yes I meant off-chain data.

[the remark this answers is not in the document]
Comment (chipshort) [reply src/part2/smart-contract-development.md:98]: In that case we don't really provide a mechanism for that. There is a very limited number of ways to get data into the contract:
- Sending an ExecuteMsg
- The chain itself provides the data in some way and the contract can query it
- Some IBC connected chain provides the data like e.g. Nois does for random data. That can then be queried through IBC.

The big problem with off-chain data is how to make sure that it is correct (i.e. the one providing it is not lying to you for their own benefit). I don't know if there is any project in Cosmos providing something like this (I heard Chainlink does on the Ethereum ecosystem).

**Review by costa2400 — COMMENTED** [review]
V1 of CosmWasm Book 2.0

**Pull request merged**